scala/scala.bzl

```diff
@@ -64,23 +66,59 @@ touch -t 198001010000 {manifest}
       progress_message="scala %s" % ctx.label,
       arguments=[])
 
-def _compile(ctx, jars, buildijar):
+def _compile(ctx, jars, dep_srcjars, buildijar):
   res_cmd = _add_resources_cmd(ctx)
   ijar_cmd = ""
   if buildijar:
     ijar_cmd = "\n{ijar} {out} {ijar_out}".format(
       ijar=ctx.file._ijar.path,
       out=ctx.outputs.jar.path,
       ijar_out=ctx.outputs.ijar.path)
+
+  sources = []
+  srcjars = []
+  for f in ctx.files.srcs:
+    #TODO this is gross but we aren't given a good "filterNot"
+    if len(_srcjar_filetype.filter([f])) == 0:
+      sources.append(f)
+    else:
+      srcjars.append(f)
+
+  all_srcjars = srcjars + list(dep_srcjars)
+
+  srcjar_cmd = "\n_SRCJAR_ARG=\"\""
+  if len(all_srcjars) > 0:
+    for srcjar in all_srcjars:
+      # Note: this is double escaped because we need to do one format call
+      # per each srcjar, but then we are going to include this in the bigger format
+      # call that is done to generate the full command
+
+      # Note: unzip has -o set (overriding files), and all of the files are unzipped into the same directory.
+      # I feel this is ok because everything should be from the same source tree, so it should have to be consistent
+      # (the whole point of bazel is resolving diamonds). That said, a good TODO might be to ensure that
+      # if there are duplicate files, they are identical (and error otherwise)
+
+      #TODO would like to be able to switch >/dev/null, -v, etc based on the user's settings
+      srcjar_cmd += """
+rm -rf {{out}}_tmp_expand_srcjars
+mkdir -p {{out}}_tmp_expand_srcjars
+unzip -o {srcjar} -d {{out}}_tmp_expand_srcjars >/dev/null
+""".format(srcjar = srcjar.path)
+    srcjar_cmd += """
+_SRCJAR_ARG=$(find {out}_tmp_expand_srcjars -type f -name "*.scala")
+"""
+
   cmd = """
 rm -rf {out}_tmp
 set -e
-mkdir -p {out}_tmp
-env JAVACMD={java} {scalac} {scala_opts} {jvm_flags} -classpath "{jars}" $@ -d {out}_tmp
+mkdir -p {out}_tmp""" + srcjar_cmd + """
+env JAVACMD={java} {scalac} {scala_opts} {jvm_flags} -classpath "{jars}" $_SRCJAR_ARG $@ -d {out}_tmp
 # Make jar file deterministic by setting the timestamp of files
+rm -rf
```

Comment:
good catch. fixed. this is from when I was moving around how the srcjar stuff was expanded